### PR TITLE
[Merged by Bors] - Builder Specs v0.2.0

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

https://github.com/sigp/lighthouse/issues/3091

Extends https://github.com/sigp/lighthouse/pull/3062, adding pre-bellatrix block support on blinded endpoints and allowing the normal proposal flow (local payload construction) on blinded endpoints. This resulted in better fallback logic because the VC will not have to switch endpoints on failure in the BN <> Builder API, the BN can just fallback immediately and without repeating block processing that it shouldn't need to. We can also keep VC fallback from the VC<>BN API's blinded endpoint to full endpoint.

## Proposed Changes

- Pre-bellatrix blocks on blinded endpoints
- Add a new `PayloadCache` to the execution layer
- Better fallback-from-builder logic

## Todos

- [x] Remove VC transition logic
- [x] Add logic to only enable builder flow after Merge transition finalization
- [x] Tests
- [x] Fix metrics
- [x] Rustdocs

### CODE REVIEW

**Comment by realbigsean** [comment]
This doesn't match the semantics of the upcoming changes to the builder API: https://github.com/flashbots/mev-boost/pull/82

But I think it'd still be worth working towards merging this as is, rather than waiting on the stabilization of that spec if we plan to use the payload cache elsewhere. This also gets us in line with the standard API spec by making sure we can provide pre-bellatrix blocks on blinded endpoints.

**Comment by paulhauner** [comment]
There's a conflict here :)

**Comment by realbigsean** [comment]
Blocked on #3070

File: common/eth2/src/lib.rs
Comment (realbigsean) [inline common/eth2/src/lib.rs:1294]: Sneaky bugfix, the spec says this should be V1, our beacon HTTP server will accept any version so this may have kept us from working cross client but wouldn't have been an issue with a full lighthouse setup

**Comment by realbigsean** [comment]
I have yet to really add docs, but this is ready for review otherwise. I am going to work on docs tomorrow. The issues linked [here](https://github.com/sigp/lighthouse/issues/3091) include a lot of the reasoning for the changes here. If this PR is too large for anyone's appetite let me know and I can work to split it up along the lines of the issues included.

File: .github/workflows/local-testnet.yml
Comment (michaelsproul) [inline .github/workflows/local-testnet.yml:43]: Is the `sleep` necessary?

File: Cargo.lock
Comment (michaelsproul) [inline Cargo.lock]: I've opened a PR upstream to see if we can remove this duplicate libp2p dependency: https://github.com/ralexstokes/ethereum-consensus/pull/139

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I don't think we need this clone

```suggestion
                        beacon_block: signed_block,
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
        let epochs_since_finalization = current_epoch.saturating_sub(
```

File: beacon_node/beacon_chain/src/chain_config.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/chain_config.rs]:
```suggestion
    /// Number of skip slots in a row before the BN refuses to use connected builders during payload construction.
```

File: beacon_node/beacon_chain/src/chain_config.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/chain_config.rs]:
```suggestion
    /// Number of skip slots in the past `SLOTS_PER_EPOCH` before the BN refuses to use connected
```

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs]: Do you mind wrapping this so it doesn't overflow 100 chars? Rustfmt usually gets confused with stuff like this

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs]: Same here

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs]: Same comment about wrapping for these logs :pray:

**Review by michaelsproul — COMMENTED** [review]
I've reviewed about 1/3 of this. Will come back for more tomorrow

**Comment by realbigsean** [comment]
@michaelsproul  I added batching in the validator registration requests in a15a0d5123ca90434fcd27476cc3affe42add0e0 because I had issues with the relay when running 2.5k validators on ropsten.

File: beacon_node/http_api/src/publish_blocks.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/publish_blocks.rs]: I think we could simplify this function a lot by using `try_into_full_block`, or the individual `From`/`Into` implementations for `Base` and `Altair` blocks.

https://github.com/sigp/lighthouse/blob/21dec6f603edd53ad8d2372941c77ab8098d3b5a/consensus/types/src/signed_beacon_block.rs#L310-L324

File: book/src/builders.md
Comment (michaelsproul) [inline book/src/builders.md]: These docs are a bit incomplete. If we tidy them up to be minimally correct (e.g. no incomplete sentences or wrong info) I think we can merge now and come back for more comprehensive docs later

File: common/eth2/src/lighthouse_vc/types.rs
Comment (michaelsproul) [inline common/eth2/src/lighthouse_vc/types.rs]: I think we need `serde(default)` on these too?

File: consensus/types/src/chain_spec.rs
Comment (michaelsproul) [inline consensus/types/src/chain_spec.rs]:
```suggestion
    fn test_domain_builder() {
```

File: validator_client/src/block_service.rs
Comment (michaelsproul) [inline validator_client/src/block_service.rs]: Could wrap this

File: validator_client/src/validator_store.rs
Comment (michaelsproul) [inline validator_client/src/validator_store.rs]:
```suggestion
        // If there is a `gas_limit` in the validator definitions yaml
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looking great on the whole, just a few minor comments and then we're ready to roll!

I've added the v2.5.0 tag too, but we can drop it from the release if we don't make the deadline

File: Cargo.lock
Comment (michaelsproul) [reply Cargo.lock]: With that PR merged I think a `cargo update` will sort this out, as `mev-rs` doesn't pin an `ethereum-consensus` version

File: beacon_node/http_api/src/publish_blocks.rs
Comment (michaelsproul) [reply beacon_node/http_api/src/publish_blocks.rs]: I was thinking something along the lines of

```rust
let full_payload = if let Ok(payload_header) = block.body().execution_payload() {
    // Logic to fetch full payload...
    Some(payload)
} else {
    None
};
block.try_into_full_block(full_payload)
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Added in ae1a500225471caac9c929daaf64195f3bea4b47

File: beacon_node/http_api/src/publish_blocks.rs
Comment (realbigsean) [reply beacon_node/http_api/src/publish_blocks.rs]: Love it! Added here: 78a1237a4aca3fab81203e3f443d4b111bcd9728

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [reply beacon_node/execution_layer/src/lib.rs]: Added line break related fixes here: 6093f992dbb4c6d313b1bad966298b5c4f172a6a

File: validator_client/src/block_service.rs
Comment (realbigsean) [reply validator_client/src/block_service.rs]: Added here: 4c67d7d625274c8c9eb9effc7f0405b3f5095c92

File: Cargo.lock
Comment (realbigsean) [reply Cargo.lock]: Looks like it worked! https://github.com/sigp/lighthouse/pull/3134/commits/5ad7632fc8481882b9367a118ae4b2f3c8c9d053#diff-13ee4b2252c9e516a0547f2891aa2105c3ca71c6d7a1e682c69be97998dfc87eL1944

**Review by michaelsproul — APPROVED** [review]
The review updates so far look good! Once you're happy with it I think we should MERGE 🐼

File: Cargo.lock
Comment (michaelsproul) [reply Cargo.lock]: Heck yeah!

File: .github/workflows/local-testnet.yml
Comment (michaelsproul) [reply .github/workflows/local-testnet.yml:43]: (don't mind if it stays)

Comment (realbigsean) [reply .github/workflows/local-testnet.yml:43]: Had to dig into this again because it's been a while since I added that. I think we do need this.

If you check out the logs in some CI runs on [other branches](https://github.com/sigp/lighthouse/runs/7460369449?check_suite_focus=true) we never get past this log line, even though the `print_logs.sh` it looks like we should be dumping the whole file. 

`Jul 22 01:12:04.334 INFO Starting from known genesis state       service: beacon`

So I think what happens is we kick off all the processes and shut them down before they are fully initialized.

File: common/eth2/src/lighthouse_vc/types.rs
Comment (realbigsean) [reply common/eth2/src/lighthouse_vc/types.rs]: I think you're right, I added that and `serde(skip_serializing_if = "Option::is_none")` to all the optional fields in that request https://github.com/sigp/lighthouse/pull/3134/commits/5f6dc08ef3e16178e41868a8618adf5e816f11a9

File: Cargo.lock
Comment (realbigsean) [reply Cargo.lock]: Ah dang there's actually a knock-on effect of this because `mev-build-rs` depends on `beacon-client-api` which used `PeerId`. I'm trying to update `beacon-client-api` here: https://github.com/ralexstokes/beacon-api-client/pull/28/files

**Comment by realbigsean** [comment]
Working on docs and resolving this: https://github.com/sigp/lighthouse/pull/3134#discussion_r933348680 

But will merge when those things are all set!

File: Cargo.lock
Comment (realbigsean) [reply Cargo.lock]: ok it looks all good now

**Comment by realbigsean** [comment]
Added docs and fixed the link check here https://github.com/sigp/lighthouse/pull/3134/commits/84e7f5375e3a41d41511593a7c17d1bc38a41488

Resolved the dependency issue here https://github.com/sigp/lighthouse/pull/3134/commits/eda5a5ca626da9042a28bc6e69d1c1132a8719d7

So I think we're good to go now :tada:

**Pull request closed without merging**
